NativeAOT-LLVM: Build for wasi sdk

yowl opened the pull request:
This PR enable a Wasi SDK build with packages (RID `wasi-wasm`).  There's a lot of changes here to cater with the new build and limitations in the Wasi SDK libc.  These limitations are around pthreads and mmap./munmap.  Otherwise its mostly infrastructure changes, some changes to the BuildIntegration scripts naturally.  Resulting wasm files can be run with wasmer, or wasmtime with the right flags (in the documentation changes in the PR).

There is no Globalization support, attempting to use it will crash.  Probably we need to build ICU with the Wasi SDK.
There is no `munmap`, attempting to do a partial unmap will result in reuse of memory and crash soon after.
There is no means to get the thread stack size.  This is interesting in that this looks to be used for at least 2 things in the runtime:

- Is the stack non null from `PalGetMaximumStackBounds`.
- Some libraries, e.g. `RegEx` use the stack size, I suppose to determine how to compile or evaluate the RegEx.

So while we could return information about the shadow stack, that might be misleading to `RegEx` as there is potentially more native stack available.  Same applies to the emscripten build I guess.

YML jobs are in for Debug and Release, but only the HelloWasm test is running (without the exceptions tests), I expect the others will mostly fail because of the lack of exceptions.

To publish
```
dotnet publish -r wasi-wasm -c Debug /p:TargetArchitecture=wasm /p:PlatformTarget=AnyCPU /p:MSBuildEnableWorkloadResolver=false --self-contained
```
And `wasmtime` can be used to run
```
wasmtime --wasm-features=threads --wasi-modules=experimental-wasi-threads bin\Debug\net8.0\wasi-wasm\native\console.wasm
```
Other runtimes may work, but we are building with pthreads (although as per the emscripten build, they wont work) so you need a runtime that has support for wasm threads.

yowl committed:
Add support for Wasi using the Wasi SDK

yowl commented:
cc @dotnet/nativeaot-llvm

SingleAccretion commented on eng/build.ps1:
```diff
@@ -291,9 +291,6 @@ if ($os -eq "browser") {
 
 if ($os -eq "wasi") {
-  # override default arch for wasi, we only support wasm
-  $arch = "wasm"
```
What's the reason for this removal?

yowl replied:
None, reinstated.

SingleAccretion commented on eng/native/configurecompiler.cmake:
```diff
@@ -241,6 +241,12 @@ elseif (CLR_CMAKE_TARGET_ARCH_WASM)
   add_definitions(-DHOST_WASM=1)
   add_definitions(-DFEATURE_64BIT_ALIGNMENT=1)
+  if (CLR_CMAKE_TARGET_OS STREQUAL wasi)
+    add_definitions(-DTARGET_WASI=1 -D_WASI_EMULATED_PROCESS_CLOCKS -D_WASI_EMULATED_SIGNAL -D_WASI_EMULATED_MMAN -D_WASI_EMULATED_GETPID)
+    # no-unsafe-buffer-usage for pal_random.c
+    # no-unused-macros for _version.c
+    add_compile_options(-Wno-deprecated-declarations -Wno-unsafe-buffer-usage -Wno-unused-macros)
+  endif ()
```
It would be more consistent to put this below (line 619).

Why is `-DTARGET_WASI=1` given code defines `TARGET_WASI` below?

yowl replied:
We have `CLR_CMAKE_TARGET_UNIX` so we don't enter the `elseif(CLR_CMAKE_TARGET_WASI)`  I could add an if inside the ``CLR_CMAKE_TARGET_UNIX`` block, just above line 619

SingleAccretion replied:
I would be biased towards doing things "the upstream way", i. e. not setting `CLR_CMAKE_TARGET_UNIX` for WASI, at least at this level.

yowl replied:
I see, so dont set `CLR_CMAKE_TARGET_UNIX` or `CLR_CMAKE_TARGET_WIN32` and in places like
```
#ifdef TARGET_UNIX
#define __stdcall
#endif
```
Where we want the UNIX definition, change to `TARGET_UNIX || TARGET_WASI` in essence ?

SingleAccretion replied:
I would avoid changing the top-level CMake configuration and do that in common code (if any), yes.

For the NAOT native runtime, if it simplifies things (i. e. leads to fewer changes), which I suspect it does, I think it is ok/good to set `TARGET_UNIX` for WASI as well.

The overall strategy is to not diverge with upstream in areas that are common (e.g. `libs.native -os wasi`) and do what makes sense for the components we own (`clr.nativeaotruntime -os wasi`).

yowl replied:
Have done as suggested not set `CLR_CMAKE_TARGET_UNIX` or `CLR_CMAKE_TARGET_WIN32` and set `TARGET_UNIX` for nativeaot

SingleAccretion commented on eng/native/configureplatform.cmake:
```diff
@@ -429,6 +429,7 @@ if(CLR_CMAKE_TARGET_OS STREQUAL emscripten)
 endif(CLR_CMAKE_TARGET_OS STREQUAL emscripten)
 
 if(CLR_CMAKE_TARGET_OS STREQUAL wasi)
+    set(CLR_CMAKE_TARGET_UNIX 1)
```
This is repository-wide configuration. Does something other than the NAOT runtime build need it?

yowl replied:
removed and set just for nativeaot (`TARGET_UNIX`)

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets:
```diff
@@ -96,11 +96,12 @@
 
-    <Delete Files="$(PublishDir)\$(TargetName).wasm" Condition="'$(_targetArchitecture)'=='wasm'"/>
-    <Copy SourceFiles="$(NativeOutputPath)$(TargetName).wasm" DestinationFolder="$(PublishDir)" Condition="'$(_targetArchitecture)'=='wasm'"/>
+    <!-- For Wasi, the .wasm file is the NativeBinaryExt -->
+    <Delete Files="$(PublishDir)\$(TargetName).wasm" Condition="'$(_targetOS)' == 'browser'"/>
+    <Copy SourceFiles="$(NativeOutputPath)$(TargetName).wasm" DestinationFolder="$(PublishDir)" Condition="'$(_targetOS)' == 'browser'"/>
```
This is confusing: comment mentions WASI, but code does `"'$(_targetOS)' == 'browser'"`.

yowl replied:
Yes, don't think we need these lines at all

yowl replied:
Removed now

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Wasi.props:
```diff
@@ -0,0 +1,19 @@
+The .NET Foundation licenses this file to you under the MIT license.
+***********************************************************************************************
+-->
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <!-- Part of workaround for lack of secondary build artifact import - https://github.com/Microsoft/msbuild/issues/2807 -->
+  <!-- required Target, but nothing to do for Wasm -->
+  <Target Name="SetupOSSpecificProps" />
+</Project>
```
I think we should rather rename the `.Browser` one to `.wasm` and use it for both. We should eventually migrate the various knobs to this file from `.Native.targets`.

yowl replied:
Done, thanks

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -118,7 +119,6 @@ The .NET Foundation licenses this file to you under the MIT license.
     <IlcCompileDependsOn Condition="'$(BuildingFrameworkLibrary)' != 'true'">Compile;ComputeIlcCompileInputs</IlcCompileDependsOn>
     <IlcCompileDependsOn Condition="'$(IlcMultiModule)' == 'true' and '$(BuildingFrameworkLibrary)' != 'true'">$(IlcCompileDependsOn);BuildFrameworkLib</IlcCompileDependsOn>
     <IlcCompileDependsOn>$(IlcCompileDependsOn);SetupOSSpecificProps</IlcCompileDependsOn>
-    <IlcCompileDependsOn>$(IlcCompileDependsOn);PrepareForILLink</IlcCompileDependsOn>
```
Presumably not intentional.

yowl replied:
oh man, I keep messing this up :-O

SingleAccretion commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System.Private.CoreLib.csproj:
```diff
@@ -30,11 +30,14 @@
+    <FeaturePortableTimer Condition="'$(TargetsUnix)' == 'true' or '$(TargetsBrowser)'=='true' or '$(TargetsWasi)'=='true'">true</FeaturePortableTimer>
+  </PropertyGroup>
+  <PropertyGroup>
+    <FeatureWasmThreads Condition="'$(TargetsWasi)' == 'true'">true</FeatureWasmThreads>
```
Am I right to understand threads are basically only to have various Unix ptthread APIs available for compilation?

If so, why does managed code need to enable them?

yowl replied:
We get this error:
```
E:\GitHub\runtimelab\src\libraries\System.Private.CoreLib\src\System\Threading\ThreadPoolBoundHandleOverlapped.cs(15,27): error CS0246: The type or namespace
 name 'PreAllocatedOverlapped' could not be found (are you missing a using directive or an assembly reference?) [E:\GitHub\runtimelab\src\coreclr\nativeaot\S
ystem.Private.CoreLib\src\System.Private.CoreLib.csproj]
E:\GitHub\runtimelab\src\libraries\System.Private.CoreLib\src\System\Threading\ThreadPoolBoundHandle.cs(216,66): error CS0246: The type or namespace name 'Pr
eAllocatedOverlapped' could not be found (are you missing a using directive or an assembly reference?) [E:\GitHub\runtimelab\src\coreclr\nativeaot\System.Pri
vate.CoreLib\src\System.Private.CoreLib.csproj]
E:\GitHub\runtimelab\src\libraries\System.Private.CoreLib\src\System\Threading\ThreadPoolBoundHandleOverlapped.cs(21,111): error CS0246: The type or namespac
e name 'PreAllocatedOverlapped' could not be found (are you missing a using directive or an assembly reference?) [E:\GitHub\runtimelab\src\coreclr\nativeaot\
System.Private.CoreLib\src\System.Private.CoreLib.csproj]
```
If we try to turn off `FeaturePortableThreadPool` we get errors like
```
E:\GitHub\runtimelab\src\libraries\System.Private.CoreLib\src\System\Threading\ThreadPoolWorkQueue.cs(1381,23): error CS0246: The type or namespace name 'Reg
isteredWaitHandle' could not be found (are you missing a using directive or an assembly reference?) [E:\GitHub\runtimelab\src\coreclr\nativeaot\System.Privat
e.CoreLib\src\System.Private.CoreLib.csproj]
```
I can work through this in more detail if you like?

SingleAccretion replied:
It would be great to get to the bottom of this. The overriding concern is that we shouldn't enable features that don't work.

yowl replied:
Threads are clearly prevalent across `System.Private.CoreLib` and not behind flags https://github.com/dotnet/runtimelab/blob/c22ff170b51086a500c378ef31eaece19a0987f7/src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems#L1129-L1213 however if we turn off the features, we can compensate with some throwing stubs in nativeaot.  There's quite a few to get past the ApiCompat check and the fact that some other parts of the code reference threads, `SafeFileHandle` being one example.

SingleAccretion commented on src/coreclr/tools/Common/JitInterface/JitConfigProvider.cs:
```diff
@@ -130,7 +130,8 @@ public string GetStringConfigValue(string name)
         private static string GetTargetSpec(TargetDetails target)
         {
-            string targetOSComponent = (target.OperatingSystem == TargetOS.Windows ? "win" : (target.OperatingSystem == TargetOS.WebAssembly ? "browser" : "unix"));
+            // Use the same Jir for browser and wasi
+            string targetOSComponent = (target.OperatingSystem == TargetOS.Windows ? "win" : (target.OperatingSystem == TargetOS.Browser || target.OperatingSystem == TargetOS.Wasi ? "browser" : "unix"));
```
I think what we want is to use the `universal` moniker for all WASM targets, same as ARM/ARM64 do, since our code generation does not depend on the target OS (and if/when it starts to, it'll be dynamically checked).

yowl replied:
Not sure what that actually does, but have made it the same, `universal`.

SingleAccretion commented on src/native/libs/System.Globalization.Native/CMakeLists.txt:
```diff
@@ -13,7 +13,7 @@ if(CLR_CMAKE_TARGET_UNIX)
     add_compile_options(-Wno-unknown-warning-option)
 
-    if (CLR_CMAKE_TARGET_UNIX_WASM)
+    if (CLR_CMAKE_TARGET_UNIX_WASM AND CLR_CMAKE_TARGET_BROWSER)
```
```suggestion
    if (CLR_CMAKE_TARGET_BROWSER)
```

SingleAccretion commented on src/tests/CMakeLists.txt:
```diff
@@ -57,6 +57,13 @@ if (CLR_CMAKE_TARGET_BROWSER)
     return()
 endif()
 
+# TODO-LLVM: WsSI cannot do the pinvoke tests
+if (CLR_CMAKE_TARGET_WASI)
+    # NativeAOT-LLVM: only compile the smoke tests. We do not (currently) need more and 'more' doesn't all compile.
+    add_subdirectory(nativeaot/SmokeTests/HelloWasm)
+    return()
+endif()
```
I think it'd be better to share it with the `if` above s.t. as we add directories we don't need to duplicate them.

Also - is the exclusion of the PI test needed given we disable the managed project. I would expect the native project to compile at least.

yowl replied:
Have moved as suggested but left out for now as it fails.  It gets the extension wrong but more fatally can't find a `main`
```
cmd.exe /C "cd . && e:\github\wasi-sdk\\bin\clang++.exe --target=wasm32-wasi-threads -fPIC -pthread -pthread -g   -shared  -o nativeaot\SmokeTests\PInvoke\libPInvokeNative.so nativeaot/SmokeTests/PInvoke/CMakeFiles/PInvokeNative.dir/PInvokeNative.cpp.obj   && cd ."
clang++: warning: argument unused during compilation: '-shared' [-Wunused-command-line-argument]
wasm-ld: error: e:/github/wasi-sdk/bin/../share/wasi-sysroot/lib/wasm32-wasi-threads/libc.a(__main_void.o): undefined symbol: main
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

SingleAccretion commented on src/tests/Common/dirs.proj:
```diff
@@ -32,6 +32,17 @@
       <DisabledProjects Include="$(TestRoot)nativeaot\SmokeTests\HelloWasm\WasmDebugging.csproj" />
     </ItemGroup>
 
+    <ItemGroup Condition="'$(TestBuildMode)' == 'nativeaot' and '$(TargetOS)' == 'wasi'">
```
Should be shared with the list above.

yowl replied:
Move, thanks

SingleAccretion commented on src/tests/Directory.Build.targets:
```diff
@@ -200,7 +200,7 @@
      <!-- There are currently no native project binaries on wasm or Android -->
      <Error  Text="The native project files are missing in $(NativeProjectOutputFolder) please run build from the root of the repo at least once"
-             Condition="'@(NativeProjectBinaries)' == '' And '$(TargetOS)' != 'browser' And '$(TargetOS)' != 'android' And '$(TargetOS)' != 'ios' And '$(TargetOS)' != 'iossimulator'"/>
+             Condition="'@(NativeProjectBinaries)' == '' And '$(TargetOS)' != 'browser' And '$(TargetOS)' != 'wasi' And '$(TargetOS)' != 'android' And '$(TargetOS)' != 'ios' And '$(TargetOS)' != 'iossimulator'"/>
```
This is not true anymore with us testing against static native libraries, but I am guessing some part of the build complains if we don't silence the error?

yowl replied:
Looks like we are ok to remove it, if things go green, locally it seems fine for wasi at least.

SingleAccretion commented on src/tests/build.cmd:
```diff
@@ -73,6 +73,7 @@ if /i "%1" == "x86"                      (set __BuildArch=x86&set processedArgs=
 if /i "%1" == "arm"                      (set __BuildArch=arm&set processedArgs=!processedArgs! %1&shift&goto Arg_Loop)
 if /i "%1" == "arm64"                    (set __BuildArch=arm64&set processedArgs=!processedArgs! %1&shift&goto Arg_Loop)
 if /i "%1" == "wasm"                     (set __BuildArch=wasm&set __TargetOS=browser&set __DistroRid=browser-wasm&set processedArgs=!processedArgs! %1&shift&goto Arg_Loop)
+if /i "%1" == "targetos"                 (set __TargetOS=%2&set __DistroRid=wasi-wasm&set processedArgs=!processedArgs! %1 %2&shift&shift&goto Arg_Loop)
```
A `targetos` argument that always sets `wasi-wasm`... Perhaps just go with a plain `wasi`?

BTW, is reason for this that something needs `__DistroRid` or that we want `TargetOS` set in test project files? We could work around the latter with dynamic checks (`OperatingSystem.IsWasi()`).

yowl replied:
If we don't set `TargetOS` it runs the `browser` tests and fails
```
:\Program Files\dotnet\sdk\8.0.100-preview.3.23178.7\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(266,5): error NETSDK1047: Assets file 'E:\GitHub\runtimelab\artifa
cts\tests\coreclr\packages\Common\test_dependencies\test_dependencies\project.assets.json' doesn't have a target for 'net8.0/browser-wasm'. Ensure that restore has run and that you have included 'n
et8.0' in the TargetFrameworks for your project. You may also need to include 'browser-wasm' in your project's RuntimeIdentifiers. [E:\GitHub\runtimelab\src\tests\nativeaot\GenerateUnmanagedEntryPo
ints\ReferencedAssembly1\ReferencedAssembly1.csproj] [E:\GitHub\runtimelab\src\tests\build.proj]
```
Without the RID looks like it defaults to `linux` somewhere:
```
C:\Program Files\dotnet\sdk\8.0.100-preview.3.23178.7\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(90,5): error NETSDK1083: The specified RuntimeIdentifier
'linux-wasm' is not recognized. [E:\GitHub\runtimelab\src\tests\Common\test_dependencies_fs\test_dependencies.fsproj] [E:\GitHub\runtimelab\src\tests\build.proj]
```

Will start with removing `targetos`.

SingleAccretion commented on src/tests/nativeaot/SmokeTests/SharedLibrary/CMakeLists.txt:
```diff
@@ -7,5 +7,7 @@ if (CLR_CMAKE_TARGET_UNIX)
 
-# add the install targets
-install (TARGETS PInvokeNative DESTINATION bin)
+if (!CLR_CMAKE_TARGET_WASI)
+  # add the install targets
+  install (TARGETS PInvokeNative DESTINATION bin)
+endif()
```
What's the reason for needing this?

yowl replied:
None, apparently reverted

SingleAccretion reviewed (commented):
Just the first pass over the build for now, will continue tomorrow.

Great work getting this off the ground!

SingleAccretion commented on docs/using-nativeaot/compiling.md:
````diff
@@ -132,6 +132,28 @@ This can be used to import WASI functions that are in other modules, either as t
+```xml
+<ItemGroup>
+  <IlcArg Include="--codegenopt:Target=wasm32-wasi-threads" />
+</ItemGroup>
````
Can we infer the right `Target` from the RID in the `.targets` instead?

yowl replied:
yes we can!

SingleAccretion commented on docs/using-nativeaot/compiling.md:
````diff
@@ -132,6 +132,28 @@ This can be used to import WASI functions that are in other modules, either as t
+
+Create the wasm with
+```
+dotnet publish -r wasi-wasm -c Debug /p:TargetArchitecture=wasm /p:PlatformTarget=AnyCPU /p:MSBuildEnableWorkloadResolver=false --self-contained```
````
Is `/p:TargetArchitecture=wasm` needed?

yowl replied:
No, removed.

SingleAccretion replied:
Looks to still be there?

SingleAccretion commented on docs/workflow/building/coreclr/nativeaot.md:
```diff
@@ -108,6 +108,13 @@ and the following two references to the project file itself:
 You should now be able to publish the project for Wasm: `dotnet publish --self-contained -r browser-wasm /p:MSBuildEnableWorkloadResolver=false`. This produces `YourApp.html` and `YourApp.js` files under `bin\<Config>\<TFM>\browser-wasm\native`. The former can be opened in the browser, the latter - run via NodeJS.
 
+### WASI
+To build with WASI, 
+- Download and install the Wasi SDK from https://github.com/WebAssembly/wasi-sdk/releases (only Windows is supported currently)
+- Set the WASI_SDK_PATH to the location where it is installed, e.g. set WASI_SDK_PATH=c:\github\wasi-sdk
+-  Run `build clr.nativeaotruntime+clr.nativeaotlibs+libs -c [Debug|Release] -a wasm -os wasi`. This will create the architecture-dependent libraries needed for linking and runtime execution, as well as the managed binaries to be used as input to ILC.
+- If you want to test the packages then they can be built with `build nativeaot.packages -a wasm -os wasi`
```
It would be better to integrate these bits into the text that talks about cloning EMSDK, e. g.:
```
- [Existing] Clone the [emsdk](https://github.com/emscripten-core/emsdk) repository...
- [New] To build for WASI, download and install the WASI SDK...
- [Existing] Run build clr.nativeaotruntime+clr.nativeaotlibs+libs -c [Debug|Release] -a wasm -os [browser|wasi]
```

yowl replied:
Done, thanks

SingleAccretion commented on eng/pipelines/runtimelab/install-icu4c.ps1:
```diff
@@ -0,0 +1,11 @@
+Invoke-WebRequest -Uri https://github.com/unicode-org/icu/releases/download/release-73-1/icu4c-73_1-src.zip -OutFile icu4c-73_1-src.zip
+
+unzip icu4c-73_1-src.zip
+
+cd icu
+
+# build the host icu first
+msbuild source\allinone\allinone.sln /p:Configuration=Debug /p:Platform=Win64 /p:SkipUWP=true
+
+# then cross compile
+
```
Is this script used?

yowl replied:
no, its a place holder for future work :-)  Shall I remove it?

SingleAccretion commented on eng/pipelines/runtimelab/install-wasi-sdk.cmd:
```diff
@@ -0,0 +1,16 @@
+if %errorlevel% NEQ 0 goto fail
+
+echo Setting WASI_SDK_PATH to %1\wasi-sdk-20.0+m
+echo ##vso[task.setvariable variable=WASI_SDK_PATH]%1\wasi-sdk-20.0+m
```
> `echo ##vso[task.setvariable variable=WASI_SDK_PATH]%1\wasi-sdk-20.0+m`

Would it be possible to not hardcode the version (not that we should go out of our way in doing so, but it'd be nice to have one place to update in CI when we upgrade).

yowl replied:
yes, ok.  Have `mv`ed the version named folder to just `wasi-sdk`  It's 2 places, but in the same file now.

SingleAccretion commented on eng/pipelines/runtimelab/runtimelab-post-build-steps.yml:
```diff
@@ -27,11 +27,18 @@ steps:
         displayName: Build host packages
 
   # Build coreclr native test output
+  # For Wasm, we are passing the osGroup where normally the archType would be so we can set the right __DistroRid in src/tests/build.cmd
```
The comment does not seem right - we are not passing `osGroup` AFAICT.

yowl replied:
Removed

SingleAccretion commented on eng/pipelines/runtimelab/runtimelab-post-build-steps.yml:
```diff
@@ -27,11 +27,18 @@ steps:
+    - script: |
+        call $(Build.SourcesDirectory)\wasm-tools\emsdk\emsdk_env
+        $(Build.SourcesDirectory)/src/tests/build$(scriptExt) nativeaot $(buildConfigUpper) ${{ parameters.archType }} targetos wasi tree nativeaot
+      displayName: Build tests
```
```suggestion
      displayName: Build WebAssembly tests
```
Consistency.

yowl replied:
Done, thanks

SingleAccretion commented on src/coreclr/gc/unix/configure.cmake:
```diff
@@ -9,6 +9,11 @@ include(CheckTypeSize)
 
 check_include_files(sys/time.h HAVE_SYS_TIME_H)
+
+if (CLR_CMAKE_TARGET_OS STREQUAL wasi)
```
`CLR_CMAKE_TARGET_OS STREQUAL wasi` -> `CLR_CMAKE_TARGET_WASI`?

yowl replied:
Done, thanks

SingleAccretion commented on src/coreclr/nativeaot/Runtime/unix/UnixContext.h:
```diff
@@ -4,7 +4,10 @@
 #ifndef __UNIX_CONTEXT_H__
 #define __UNIX_CONTEXT_H__
 
+// WASI has no ucontext.h file - TODO-LLVM: create and include one?
```
```suggestion
// WASM has no thread state contexts.
```
Probably could be `#ifdef`ed for `TARGET_WASM` at large.

yowl replied:
Thanks, have `#ifdef` ed out the whole file.  Had to remove `GcScanRootsWorker` instead but that's ok as doesn't apply to Wasm anyway.

SingleAccretion commented on src/coreclr/nativeaot/Runtime/wasm/pal_compiler.h:
```diff
@@ -0,0 +1,74 @@
+
+#pragma once
+
+#ifndef __cplusplus
+#include <stdbool.h>
+#endif
```
I'd like to understand the purpose of these `pal_` files. How much of them are needed? Are they supposed to be exact copies of some other files in the tree? If not - stuff like `#if defined(TARGET_ANDROID)` code could presumably be pruned.

yowl replied:
Think they were part of the journey that I didn't clear up.  Have removed them now (they were copies from the Globalization folder)

yowl committed:
Update src/native/libs/System.Globalization.Native/CMakeLists.txt

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

yowl committed:
Address feedback part 1

Turn off CLR_CMAKE_TARGET_WIN32 and CLR_CMAKE_TARGET_UNIX for WASI
Add thread stubs and turn off thread features
Tidy up some test harness changes

yowl committed:
Merge branch 'wasi-sdk' of https://github.com/yowl/runtimelab into wasi-sdk

yowl committed:
Address feedback part 1

Turn off CLR_CMAKE_TARGET_WIN32 and CLR_CMAKE_TARGET_UNIX for WASI
Add thread stubs and turn off thread features
Tidy up some test harness changes

yowl committed:
Address feedback part 2

yowl committed:
Merge remote-tracking branch 'origin/feature/NativeAOT-LLVM' into wasi-sdk

# Conflicts:
#	src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets

yowl committed:
correct property rename

dont show errors where not applicable
exclude UnixContext for all Wasm

yowl committed:
sync run script with build

yowl committed:
remove unused and incomplete script.

yowl commented:
Outstanding question is just for [src/tests/build.cmd](https://github.com/dotnet/runtimelab/pull/2268/files/cdb638c92681154163d84b6f4539974335341c67#diff-bde9427e24c7e6de4ccf97b52fdfa61b5b7403f40b26692f1ee5bd07044e69d3)  I think.  Thanks!

SingleAccretion commented on docs/workflow/building/coreclr/nativeaot.md:
````diff
@@ -70,7 +70,8 @@ For the runtime libraries:
   ./emsdk activate 3.1.23
   ```
-- Run `build clr.nativeaotruntime+clr.nativeaotlibs+libs -c [Debug|Release] -a wasm -os browser`. This will create the architecture-dependent libraries needed for linking and runtime execution, as well as the managed binaries to be used as input to ILC.
+- To build for WASI, download and install the Wasi SDK from https://github.com/WebAssembly/wasi-sdk/releases (only Windows is supported currently) and set the `WASI_SDK_PATH` to the location where it is installed, e.g. `set WASI_SDK_PATH=c:\github\wasi-sdk`
````
```suggestion
- To build for WASI, download and install the Wasi SDK from https://github.com/WebAssembly/wasi-sdk/releases (only Windows is supported currently) and set the `WASI_SDK_PATH` environment variable to the location where it is installed, e.g. `set WASI_SDK_PATH=c:\github\wasi-sdk`.
```

SingleAccretion commented on docs/workflow/building/coreclr/nativeaot.md:
```diff
@@ -108,6 +109,13 @@ and the following two references to the project file itself:
+- Set the WASI_SDK_PATH to the location where it is installed, e.g. set WASI_SDK_PATH=c:\github\wasi-sdk
+-  Run `build clr.nativeaotruntime+clr.nativeaotlibs+libs -c [Debug|Release] -a wasm -os wasi`. This will create the architecture-dependent libraries needed for linking and runtime execution, as well as the managed binaries to be used as input to ILC.
+- If you want to test the packages then they can be built with `build nativeaot.packages -a wasm -os wasi`
+ 
```
```suggestion
```
Now integrated into the general instructions.

SingleAccretion commented on eng/build.ps1:
```diff
@@ -291,7 +291,7 @@ if ($os -eq "browser") {
 
 if ($os -eq "wasi") {
-  # override default arch for wasi, we only support wasm
+  # override default arch for Browser, we only support wasm
```
```suggestion
  # override default arch for wasi, we only support wasm
```
Reducing the diff.

SingleAccretion commented on eng/native/configureplatform.cmake:
```diff
@@ -461,7 +461,9 @@ if(CLR_CMAKE_TARGET_UNIX)
     endif()
 else()
-    set(CLR_CMAKE_TARGET_WIN32 1)
+    if(NOT (CLR_CMAKE_TARGET_WASI))
+        set(CLR_CMAKE_TARGET_WIN32 1)
+    endif()
```
This fix looks upstreamable.

yowl replied:
https://github.com/dotnet/runtime/pull/86208

SingleAccretion commented on src/coreclr/gc/unix/configure.cmake:
```diff
@@ -9,6 +9,11 @@ include(CheckTypeSize)
 check_include_files(sys/time.h HAVE_SYS_TIME_H)
+
+if (CLR_CMAKE_TARGET_WASI)
+  set(CMAKE_REQUIRED_DEFINITIONS "-D_WASI_EMULATED_MMAN")
```
This is set both here and in `configurecompiler.cmake`. Does it need to be in both (I presume the NAOT runtime needs it, not `libs.native`)?

yowl replied:
Yes, its the runtime.  Looks like we need both.  The one you highlight here looks like it is so that in the configure phase it finds a `sys/mman.h` for `check_include_files(sys/mman.h HAVE_SYS_MMAN_H)`  If it doesn't then it fails with 
```
e:\github\wasi-sdk\\bin\clang++.exe --target=wasm32-wasi-threads -DBUILDENV_DEBUG=1 -DDEBUG -DFEATURE_64BIT_ALIGNMENT -DFEATURE_64BIT_ALIGNMENT=1 -DFEATURE_BASICFREEZE -DFEATURE_CACHED_INTERFACE_
  DISPATCH -DFEATURE_CONSERVATIVE_GC -DFEATURE_CUSTOM_IMPORTS -DFEATURE_DYNAMIC_CODE -DFEATURE_GC_STRESS -DFEATURE_NATIVEAOT -DFEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP -DHOST_WASM=1 -DNATIVEAOT
   -DNO_UI_ASSERT -DPAL_STDCPP_COMPAT -DTARGET_32BIT -DTARGET_UNIX -DTARGET_WASI -DTARGET_WASM -DURTBLDENV_FRIENDLY=Debug -DUSE_PORTABLE_HELPERS -DVERIFY_HEAP -D_DBG -D_DEBUG -D_FILE_OFFSET_BITS=64
   -D_LIB -D_LIBUNWIND_DISABLE_ZERO_COST_APIS=1 -D_LIBUNWIND_IS_NATIVE_ONLY -D_WASI_EMULATED_GETPID -D_WASI_EMULATED_MMAN -D_WASI_EMULATED_PROCESS_CLOCKS -D_WASI_EMULATED_SIGNAL -IE:/GitHub/runtime
  lab/src/native -IE:/GitHub/runtimelab/src/coreclr/pal/prebuilt/inc -IE:/GitHub/runtimelab/artifacts/obj -IE:/GitHub/runtimelab/src/coreclr/nativeaot/Runtime/inc -IE:/GitHub/runtimelab/src/coreclr
  /nativeaot/Runtime/. -IE:/GitHub/runtimelab/src/coreclr/nativeaot/Runtime/../../gc -IE:/GitHub/runtimelab/src/coreclr/nativeaot/Runtime/../../gc/env -IE:/GitHub/runtimelab/src/coreclr/nativeaot/R
  untime/unix -IE:/GitHub/runtimelab/src/coreclr/nativeaot/Runtime/../../pal/inc/rt -IE:/GitHub/runtimelab/src/coreclr/nativeaot/Runtime/wasm -IE:/GitHub/runtimelab/artifacts/obj/coreclr/wasi.wasm.
  Debug/nativeaot/Runtime -pthread -g -fms-extensions -Wno-unused-variable -Wno-unused-value -Wno-unused-function -Wno-tautological-compare -Wno-invalid-offsetof -Wno-deprecated-declarations -Wno-u
  nsafe-buffer-usage -Wno-unused-macros -std=gnu++11 -MD -MT nativeaot/Runtime/Portable/CMakeFiles/PortableRuntime.dir/__/__/__/gc/unix/gcenv.unix.cpp.obj -MF nativeaot\Runtime\Portable\CMakeFiles\
  PortableRuntime.dir\__\__\__\gc\unix\gcenv.unix.cpp.obj.d -o nativeaot/Runtime/Portable/CMakeFiles/PortableRuntime.dir/__/__/__/gc/unix/gcenv.unix.cpp.obj -c E:/GitHub/runtimelab/src/coreclr/gc/u
  nix/gcenv.unix.cpp
  E:/GitHub/runtimelab/src/coreclr/gc/unix/gcenv.unix.cpp:53:3: error: "sys/mman.h required by GC PAL"
   #error "sys/mman.h required by GC PAL"
```
Even though the compilation includes `-D_WASI_EMULATED_MMAN`

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets:
```diff
@@ -96,11 +96,8 @@
-    <Delete Files="$(PublishDir)\$(TargetName).wasm" Condition="'$(_targetArchitecture)'=='wasm'"/>
-    <Copy SourceFiles="$(NativeOutputPath)$(TargetName).wasm" DestinationFolder="$(PublishDir)" Condition="'$(_targetArchitecture)'=='wasm'"/>
+    <Delete Files="$(PublishDir)\$(TargetName).js" Condition="'$(_targetOS)' == 'browser'"/>
+    <Copy SourceFiles="$(NativeOutputPath)$(TargetName).js" DestinationFolder="$(PublishDir)" Condition="'$(_targetOS)' == 'browser'"/>
```
Does it still copy the `.wasm` when targeting `browser-wasm`?

yowl replied:
It didn't but does now, thanks

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -339,23 +341,37 @@ The .NET Foundation licenses this file to you under the MIT license.
-             Condition="'$(EMSDK)' == ''" />
+   <PropertyGroup Condition="'$(_targetOS)' == 'wasi'">
+      <CompileWasmArgs>-fvisibility=default -mllvm -combiner-global-alias-analysis=false -mllvm -disable-lsr --sysroot=&quot;$(WASI_SDK_PATH)/share/wasi-sysroot&quot; -pthread -target wasm32-wasi-threads -c</CompileWasmArgs>
+      <CompileWasmArgs Condition="'$(NativeDebugSymbols)' != 'true'">$(CompileWasmArgs) -O2</CompileWasmArgs>
```
Presumably this should use the recently introduced `EmccOptimizationSetting`, though that would be overloading things a bit (in the sense that is `clang++ -O3` != `emcc -O3`).

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -339,23 +341,37 @@ The .NET Foundation licenses this file to you under the MIT license.
+
+      <ExeExt Condition="'$(OS)' == 'Windows_NT'">.exe</ExeExt>
+      <!-- using Emscripten's clang++ because of a crash in wasi-sdk's clang++ (https://github.com/WebAssembly/wasi-sdk/issues/326) -->
+      <WasmCompilerPath>&quot;$(EMSDK)/upstream/bin/clang++$(ExeExt)&quot;</WasmCompilerPath>
+      <WasmLinkerPath>&quot;$(WASI_SDK_PATH)/bin/clang&quot;</WasmLinkerPath>
```
Meta question for my education: if not the clang crash - could we build for WASI without EMSDK?

yowl replied:
Yes, it runs out of memory supposedly.  I have 64GB and in task manager it move s when compiling but not much, it could spike within task managers sample period, so I might not be seeing it.  Wasi SDK don't use tip of tree for LLVM so we aren't expecting a new version until v17.  I will raise an issue with LLVM as there are other differences with emscripten, so the bug may still be there.

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -339,23 +341,37 @@ The .NET Foundation licenses this file to you under the MIT license.
+    </PropertyGroup>
+
+    <Exec Command="$(WasmCompilerPath) &quot;%(LlvmObjects.Identity)&quot; -o &quot;%(LlvmObjects.NativeObject)&quot; $(CompileWasmArgs)" Condition="'$(EMSDK)' != ''" />
+    <Message Text="Emscripten not found, not linking WebAssembly for Wasi. To enable WebAssembly linking, install Emscripten and ensure the EMSDK environment variable points to the directory containing upstream/emscripten/emcc.bat"
```
```suggestion
    <Message Text="Emscripten not found, not linking WebAssembly. To enable WebAssembly linking, install Emscripten and ensure the EMSDK environment variable points to the directory containing upstream/emscripten/emcc.bat"
```
This is under `'$(_targetOS)' == 'browser'`.

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -425,15 +441,24 @@ The .NET Foundation licenses this file to you under the MIT license.
+      <CustomLinkerArg Include="@(NativeLibrary->'&quot;%(Identity)&quot;'->Replace(&quot;\&quot;, &quot;/&quot;))" />
+      <CustomLinkerArg Include="@(NativeObjects->'&quot;%(Identity)&quot;'->Replace(&quot;\&quot;, &quot;/&quot;))" />
+      <CustomLinkerArg Include="-o &quot;$(NativeBinary.Replace(&quot;\&quot;, &quot;/&quot;))&quot;" />
+      <CustomLinkerArg Condition="'$(NativeDebugSymbols)' != 'true'" Include="-O2 -flto" />
```
Same `EmccOptimizationSetting` question.

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -425,15 +441,24 @@ The .NET Foundation licenses this file to you under the MIT license.
+      <CustomLinkerArg Condition="'$(NativeDebugSymbols)' != 'true'" Include="-O2 -flto" />
+      <CustomLinkerArg Condition="'$(NativeDebugSymbols)' == 'true'" Include="-g3" />
+      <CustomLinkerArg Condition="$(NativeLib) == 'Shared'" Include="-Wl,--export,NativeAOT_StaticInitialization" />
+      <CustomLinkerArg Include="@(LinkerArg)" />
```
There are a few things duplicated here (e. g. the `"-Wl,--export,NativeAOT_StaticInitialization"`). Can we share them instead?

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -446,11 +471,22 @@ The .NET Foundation licenses this file to you under the MIT license.
+    <ItemGroup Condition  ="'$(_targetOS)' == 'wasi'" >
+      <!-- Wasi has lots of undefined symbols currently, mostly globalization -->
+      <CustomLinkerArg Include="-Wl,--unresolved-symbols=ignore-all -lstdc++ -Wl,--error-limit=0 -lpthread -pthread" />
+      <CustomLinkerArg Include="--sysroot=&quot;$(WASI_SDK_PATH.Replace(&quot;\&quot;, &quot;/&quot;))/share/wasi-sysroot&quot; -target wasm32-wasi-threads" />
```
> `wasm32-wasi-threads`

It would be good to introduce a variable (`IlcLlvmTarget`) and use it here and above when passing `--codegenopt` to ILC.

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -446,11 +471,22 @@ The .NET Foundation licenses this file to you under the MIT license.
-             Condition="'$(NativeCodeGen)' == 'llvm' and '$(EMSDK)' == ''" Importance="High" />
+             Condition="'$(NativeCodeGen)' == 'llvm' and '$(_targetOS)' == 'browser'and '$(EMSDK)' == ''" Importance="High" />
+    <Message Text="Wasi SDK not found, not linking WebAssembly. To enable WebAssembly linking, install the Wasi SDK and ensure the WASI_SDK_PATH environment variable points to the directory containing bin/clang"
+             Condition="'$(NativeCodeGen)' == 'llvm' and '$(_targetOS)' == 'wasi'and '$(WASI_SDK_PATH)' == ''" Importance="High" />
```
The message here and above give conflicting guidance:
```
... WASI_SDK_PATH environment variable points to the directory containing share/wasi-sysroot
```
```
... WASI_SDK_PATH environment variable points to the directory containing bin/clang
```

yowl replied:
Now both `share/wasi-sysroot`

SingleAccretion commented on src/coreclr/nativeaot/Runtime/wasm/wasi.cpp:
```diff
@@ -0,0 +1,63 @@
+
+// TODO-LLVM: replace with ICU
+
+static int GlobalizationNative_IndexOf(void* pSortHandle,
```
Used?

yowl replied:
No, deleted

SingleAccretion commented on src/coreclr/nativeaot/Runtime/wasm/wasi.c:
```diff
@@ -0,0 +1,102 @@
+#include "wasi.h"
+#include <pthread.h>
+
+// copied from internal/pthread_impl.h
+#define __SU (sizeof(size_t)/sizeof(int))
+
+#define _a_stacksize __u.__s[0]
+#define _a_guardsize __u.__s[1]
+#define _a_stackaddr __u.__s[2]
+#define _a_detach __u.__i[3*__SU+0]
+
+struct __pthread {
```
How much of this is needed (everything)?

Am I right to understand that we are basically polyfilling some pthread here because the underlying implementation doesn't provide them?

yowl replied:
Yes, Wasi_SDK's libc does not provide the non portable pthread methods, for us the one we want is `pthread_getattr_np` so I copied basically the emscripten version, but I need access to the private fields of the __pthread so I copied the definition of the struct and a few other macros from the internal file in the Wasi SDK.  It's all a bit fragile, I know.

SingleAccretion replied:
It is indeed fragile.

I dug around a bit and found these release notes:

https://github.com/WebAssembly/wasi-libc/blob/a6f871343313220b76009827ed0153586361c0d5/libc-top-half/musl/WHATSNEW#L873

Perhaps we could set `_GNU_SOURCE` and things would work?

SingleAccretion commented on src/coreclr/nativeaot/Runtime/wasm/wasi.cpp:
```diff
@@ -0,0 +1,63 @@
+  switch(resource_id)
+  {
+    case RLIMIT_AS :
+      // TODO-LLVM: how to implement rlimits for Wasi
+      // The gc wants to know this in GetRestrictedPhysicalMemoryLimit.  Will try with an arbritrary value of 1GB
+      ret_rlimit->rlim_cur = 1024 * 1024 * 1024;
+      ret_rlimit->rlim_max = 1024 * 1024 * 1024;
```
How is the Browser path implemented for this?

yowl replied:
Emscripten's muslc implements it with syscalls `SYS_prlimit64` and `SYS_getrlimit` : https://github.com/emscripten-core/emscripten/blob/90d76db21200d42f5536bc3a01a22303e398d8b4/system/lib/libc/musl/src/misc/getrlimit.c#L7-L24

Wasi SDK's does not include that file, https://github.com/WebAssembly/wasi-libc/blob/1dfe5c302d1c5ab621f7abf04620fae92700fd22/Makefile#L88-L101

Let me check the implementation of the syscalls:

Cut and paste the relevant bits, its more generous than what this PR has.  Shall I change the limits to `~0ULL` ?
```c
#define RLIM_INFINITY (~0ULL)

int __syscall_prlimit64(int pid, int resource, intptr_t new_limit, intptr_t old_limit) {
  REPORT(prlimit64);
  struct rlimit *old = (struct rlimit *)old_limit;
  if (old) { // just report no limits
    old->rlim_cur = RLIM_INFINITY;
    old->rlim_max = RLIM_INFINITY;
  }
  return 0;
}

int __syscall_ugetrlimit(int resource, intptr_t rlim) {
  REPORT(ugetrlimit);
  struct rlimit * limits = (struct rlimit *)rlim;
  limits->rlim_cur = RLIM_INFINITY;
  limits->rlim_max = RLIM_INFINITY;
  return 0; // just report no limits
}
```

SingleAccretion replied:
> Shall I change the limits to ~0ULL ?

Seems reasonable.

SingleAccretion commented on src/coreclr/nativeaot/Runtime/CMakeLists.txt:
```diff
@@ -153,11 +153,17 @@ else()
     )
+    if (CLR_CMAKE_TARGET_WASI)
+      list(APPEND COMMON_RUNTIME_SOURCES
+        wasm/wasi.cpp
+        wasm/wasi.c
```
Are two files important (can we have one/one for each logical polyfill area)?

yowl replied:
Have separated into logical areas

yowl committed:
Update docs/workflow/building/coreclr/nativeaot.md

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

yowl committed:
Update eng/build.ps1

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

yowl committed:
Update docs/workflow/building/coreclr/nativeaot.md

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

yowl committed:
Update src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

yowl committed:
address feedback, tidy up polyfills, build integration, docs

yowl committed:
Merge branch 'wasi-sdk' of https://github.com/yowl/runtimelab into wasi-sdk

SingleAccretion commented on docs/using-nativeaot/compiling.md:
````diff
@@ -132,6 +132,23 @@ This can be used to import WASI functions that are in other modules, either as t
+
+Create the wasm with
+```
+dotnet publish -r wasi-wasm -c Debug /p:PlatformTarget=AnyCPU /p:MSBuildEnableWorkloadResolver=false --self-contained```
````
The triple ` should be moved to a new line, otherwise the rendering is not right:

![image](https://github.com/dotnet/runtimelab/assets/62474226/eea50074-4d96-44cb-951f-48f5d46a060c)

yowl replied:
Thanks, done

![image](https://github.com/dotnet/runtimelab/assets/2720041/d4d48201-f542-4b9f-8f1f-ef946dbb1e36)

SingleAccretion commented on docs/using-nativeaot/compiling.md:
````diff
@@ -132,6 +132,23 @@ This can be used to import WASI functions that are in other modules, either as t
+Create the wasm with
+```
+dotnet publish -r wasi-wasm -c Debug /p:PlatformTarget=AnyCPU /p:MSBuildEnableWorkloadResolver=false --self-contained```
+`wasmtime` is the only tested runtime.  An example invocation with wasmer:
````
I would put a note about only `wasmtime` being tested at the end (with the `Note that while ...`), and switch the order in which they are mentioned s.t. we direct to the more tested experience first.

yowl replied:
Done, actually it is wasmer that is tested.  (I started with wasmtime for the tests and switched to wasmer.)

SingleAccretion commented on docs/using-nativeaot/compiling.md:
````diff
@@ -132,6 +132,23 @@ This can be used to import WASI functions that are in other modules, either as t
+```
+wasmtime --wasm-features=threads --wasi-modules=experimental-wasi-threads bin\Debug\net8.0\wasi-wasm\publish\console.wasm
+```
+Note that while we are enabling threads, they do not work attempting to create a thread will most likely crash.
````
Will it crash or throw PNSE (I would expect the latter)?

yowl replied:
It crashes, because we don't have an implementation to get the stack trace
```
error: failed to run `bin\Debug\net8.0\wasi-wasm\publish\console.wasm`
╰─▶ 1: RuntimeError: unreachable
           at undefined_stub (console.wasm[32]:0xcb9c)
           at S_P_CoreLib_System_Diagnostics_StackTrace__emscripten_get_callstack (console.wasm[4884]:0x1c88c8)
           at S_P_CoreLib_System_Diagnostics_StackTrace__InitializeForCurrentThread (console.wasm[4648]:0x19440a)
           at S_P_CoreLib_System_Diagnostics_StackTrace___ctor_1 (console.wasm[4400]:0x15fcd0)
           at S_P_CoreLib_System_Exception__DispatchExLLVM (console.wasm[4397]:0x15f844)
           at S_P_CoreLib_System_Runtime_EH__RhpThrowEx (console.wasm[4396]:0x15f6ee)
           at RhpThrowEx (console.wasm[2121]:0x26af1)
           at S_P_CoreLib_System_Threading_Thread__ThrowIfNoThreadStart (console.wasm[4736]:0x1a6098)
           at S_P_CoreLib_System_Threading_Thread__Start_2 (console.wasm[4190]:0x13d367)
           at S_P_CoreLib_System_Threading_Thread__Start_1 (console.wasm[4180]:0x13c1d2)
           at console_Program___Main__ (console.wasm[4177]:0x13b9bd)
           at console__Module___MainMethodWrapper (console.wasm[4509]:0x171657)
           at console__Module___StartupCodeMain (console.wasm[4498]:0x1709c8)
           at main (console.wasm[18208]:0x10e9864)
           at __main_void (console.wasm[18211]:0x10e99ae)
           at _start (console.wasm[48]:0xcc81)
```

SingleAccretion replied:
Kay, so it crashes because it cannot throw PNSE \:).

SingleAccretion commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -425,32 +444,48 @@ The .NET Foundation licenses this file to you under the MIT license.
+      <CustomLinkerArg Condition="'$(Optimize)' != 'true'" Include="$(WasmOptimizationSetting) -flto" />
+      <CustomLinkerArg Condition="$(NativeLib) == 'Shared'" Include="-Wl,--export,NativeAOT_StaticInitialization" />
+      <CustomLinkerArg Condition="'$(IlcLlvmTarget)' != ''" Include="-target $(IlcLlvmTarget)" />
+      <CustomLinkerArg Include="@(LinkerArg)" />
```
We should always add user settings (`Include="@(LinkerArg)"`) last. Linker command lines allow one to override earlier options with later ones (I believe `wasm-ld` follows this as well, not sure about `emcc`). In any case, it makes debugging easier.

yowl replied:
Done.

SingleAccretion commented on src/coreclr/nativeaot/Runtime/CMakeLists.txt:
```diff
@@ -153,11 +153,17 @@ else()
+        wasm/wasi_pthread.c
+        wasm/wasi_mmap.cpp
+      )
+    endif()
```
This section is for GC files. It would look better to move them down to the clause where we include other wasm-specific files.

yowl replied:
Moved, thanks

SingleAccretion commented on src/coreclr/nativeaot/Runtime/wasm/wasi.h:
```diff
@@ -0,0 +1,42 @@
+                                              int cwSourceLength,
+                                              int options,
+                                              void* pMatchedLength);
+*/
```
```suggestion
```

yowl replied:
Removed, thanks

SingleAccretion commented on src/coreclr/nativeaot/Runtime/wasm/wasi_mmap.cpp:
```diff
@@ -0,0 +1,27 @@
+#include <string.h>
+
+#include "CommonTypes.h"
+#include "../CommonMacros.h"
```
```suggestion
#include "CommonMacros.h"
```
I do believe this should work...

yowl replied:
Indeed, thanks

SingleAccretion commented on src/coreclr/nativeaot/Runtime/wasm/wasi_mmap.cpp:
```diff
@@ -0,0 +1,27 @@
+#include "CommonTypes.h"
+#include "../CommonMacros.h"
+
+void *mmap_wasi(void *addr, size_t length, int prot, int flags,
```
We should have a comment explaining why this exists.

yowl replied:
Added.

SingleAccretion commented on src/native/corehost/common.cmake:
```diff
@@ -19,7 +19,7 @@ if (CLR_CMAKE_TARGET_WIN32)
         list(APPEND RESOURCES ${CMAKE_CURRENT_LIST_DIR}/native.rc)
     endif()
-else()
+elseif(CLR_CMAKE_TARGET_UNIT)
     list(APPEND SOURCES ${VERSION_FILE_PATH})
 endif()
```
> `CLR_CMAKE_TARGET_UNIT`

Looks like a typo.

yowl replied:
Thanks

SingleAccretion commented on src/native/libs/CMakeLists.txt:
```diff
@@ -54,7 +54,10 @@ if (CLR_CMAKE_TARGET_UNIX OR CLR_CMAKE_TARGET_BROWSER OR CLR_CMAKE_TARGET_WASI)
 
-        add_subdirectory(System.Globalization.Native)
+        if (CLR_CMAKE_TARGET_BROWSER)
+            # no ICU for wasi
```
```suggestion
            # NativeAOT-LLVM: no ICU for WASI yet.
```

SingleAccretion commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System.Private.CoreLib.csproj:
```diff
@@ -331,6 +331,9 @@
   <ItemGroup Condition="'$(TargetsBrowser)'=='true' and '$(FeatureWasmThreads)' != 'true'">
     <Compile Include="System\Threading\PreAllocatedOverlapped.NativeAot.Wasm.cs" />
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetsWasi)'=='true'">
+    <Compile Include="System\Threading\ThreadApiCompat.Wasi.cs" />
+  </ItemGroup>
```
So, when targeting `Browser`, we too don't support threading and all, but this problem doesn't come up. Why?

yowl replied:
Because we turn off `FeaturePortableThreadPool` for Wasi so we don't get , for example, `ThreadPool.Portable.cs`

SingleAccretion reviewed (commented):
There a larger question at play in that we are basically enabling threading to get things to compile. I wonder if we will need to invest into some "singlethreaded" pthread polyfill to have a single-threaded WASI implementation (or -- perhaps -- we say that's simply not going to be supported and we need to get to fixing #2148 sooner rather than later).

yowl committed:
remove pthread polyfill

add Wasi _GNU_SOURCES_

yowl committed:
reinclude implementation as WASI SDK does not.

yowl commented:
We can set `_GNU_SOURCE` and remove our own prototype for `pthread_getattr_np` but the WASI SDK Makefile does not include an implementation, so we still need that.  pthreads are a WIP for WASI SDK.

yowl committed:
more feedback

yowl committed:
Update src/native/libs/CMakeLists.txt

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

yowl commented:
> we say that's simply not going to be supported and we need to get to fixing #2148 sooner rather than later

I would prefer this.  Part of my personal motivation for Wasi-SDK is to do some work on wit-bindgen for NativeAOT-LLVM where at the moment at least I don't need threads or exceptions, I just want to get the bindings correct.

SingleAccretion commented on src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:
```diff
@@ -1184,6 +1193,9 @@ REDHAWK_PALEXPORT bool PalGetMaximumStackBounds(_Out_ void** ppStackLowOut, _Out
         status = pthread_attr_get_np(thread, &attr);
 #elif HAVE_PTHREAD_GETATTR_NP
         status = pthread_getattr_np(thread, &attr);
+#elif TARGET_WASI
+        // implementation provided in wasi.c
+        status = pthread_getattr_np(thread, &attr);
```
Should not be needed with `_GNU_SOURCE`?

yowl replied:
Hmm, apparently we still need this one:
```
e:\github\wasi-sdk\\bin\clang++.exe --target=wasm32-wasi-threads -DBUILDENV_DEBUG=1 -DDEBUG -DFEATURE_64BIT_ALIGNMENT -DFEATURE_64BIT_ALIGNMENT=1 -DFEATURE_BASICFREEZE -DFEATURE_CACHED_INTERFACE_
  DISPATCH -DFEATURE_CONSERVATIVE_GC -DFEATURE_CUSTOM_IMPORTS -DFEATURE_DYNAMIC_CODE -DFEATURE_GC_STRESS -DFEATURE_NATIVEAOT -DFEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP -DHOST_WASM=1 -DNATIVEAOT
   -DNO_UI_ASSERT -DPAL_STDCPP_COMPAT -DTARGET_32BIT -DTARGET_UNIX -DTARGET_WASI -DTARGET_WASM -DURTBLDENV_FRIENDLY=Debug -DUSE_PORTABLE_HELPERS -DVERIFY_HEAP -D_DBG -D_DEBUG -D_FILE_OFFSET_BITS=64
   -D_GNU_SOURCE -D_LIB -D_LIBUNWIND_DISABLE_ZERO_COST_APIS=1 -D_LIBUNWIND_IS_NATIVE_ONLY -D_WASI_EMULATED_GETPID -D_WASI_EMULATED_MMAN -D_WASI_EMULATED_PROCESS_CLOCKS -D_WASI_EMULATED_SIGNAL -IE:/
  GitHub/runtimelab/src/native -IE:/GitHub/runtimelab/src/coreclr/pal/prebuilt/inc -IE:/GitHub/runtimelab/artifacts/obj -IE:/GitHub/runtimelab/src/coreclr/nativeaot/Runtime/inc -IE:/GitHub/runtimel
  ab/src/coreclr/nativeaot/Runtime/. -IE:/GitHub/runtimelab/src/coreclr/nativeaot/Runtime/../../gc -IE:/GitHub/runtimelab/src/coreclr/nativeaot/Runtime/../../gc/env -IE:/GitHub/runtimelab/src/corec
  lr/nativeaot/Runtime/unix -IE:/GitHub/runtimelab/src/coreclr/nativeaot/Runtime/../../pal/inc/rt -IE:/GitHub/runtimelab/src/coreclr/nativeaot/Runtime/wasm -IE:/GitHub/runtimelab/artifacts/obj/core
  clr/wasi.wasm.Debug/nativeaot/Runtime -pthread -pthread -g -fms-extensions -Wno-unused-variable -Wno-unused-value -Wno-unused-function -Wno-tautological-compare -Wno-invalid-offsetof -Wno-depreca
  ted-declarations -Wno-unsafe-buffer-usage -Wno-unused-macros -std=gnu++11 -MD -MT nativeaot/Runtime/Portable/CMakeFiles/PortableRuntime.dir/__/unix/PalRedhawkUnix.cpp.obj -MF nativeaot\Runtime\Po
  rtable\CMakeFiles\PortableRuntime.dir\__\unix\PalRedhawkUnix.cpp.obj.d -o nativeaot/Runtime/Portable/CMakeFiles/PortableRuntime.dir/__/unix/PalRedhawkUnix.cpp.obj -c E:/GitHub/runtimelab/src/core
  clr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp
  E:/GitHub/runtimelab/src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:1197:2: error: Dont know how to get thread attributes on this platform!

  #error Dont know how to get thread attributes on this platform!
```

yowl replied:
Looks like the configure phase would need need `_GNU_SOURCE`.  Will try that

yowl replied:
Although we can get the defintion with `_GNU_SOURCE`, we still have no implementation, as the build of libc is appararently without `_GNU_SOURCE`.  I  dumped the symbols with `llvm-nm` and searched :-(
![image](https://github.com/dotnet/runtimelab/assets/2720041/0d0f54d9-9118-4417-aa0e-70dd60be48f4)

SingleAccretion reviewed (approved):
LGTM with a few comments. Thank you for addressing the feedback!

It is very unfortunate that we have to add the pthread hack, hopefully we'll get rid of it (as it's just for stack bounds, we can get those in other ways).

yowl commented:
`_GNU_SOURCE` allowed us to remove the `getrlimit` function, `wasi.cpp` no longer exists.

SingleAccretion commented:
Remove the declarations as well then?

yowl commented:
> Remove the declarations as well then?

Not sure I really understand how WASI are setting this all up.  The declarations are behind `__wasilibc_unmodified_upstream`  which doesn't seem like we should be setting it.  If we set it we need a `bits/errno.h` which they don't provide.

Edit.  I see  I already asked about this and the response was:

The immediate purpose is to comment out parts of musl which aren't supported. WASI doesn't currently have a getrlimit function, so we comment out the rlimit struct.

Further edit.  I assume `GetPhysicalMemoryLimit` returns no limit and `getrlimit` is never called so we get away with having no implementation, but need the definition for compilation only.

SingleAccretion commented:
I see; funky. Let's leave it as-is then.

yowl committed:
Merge remote-tracking branch 'origin/feature/NativeAOT-LLVM' into wasi-sdk

# Conflicts:
#	docs/workflow/building/coreclr/nativeaot.md
#	eng/pipelines/runtimelab.yml

yowl commented:
@jkotas, There has been some discussion since this was approved, but nothing was changed.  So unless @SingleAccretion has further comments, I think this is ok to go.

Many thanks for the review.

jkotas commented on src/coreclr/nativeaot/Runtime/wasm/wasi_pthread.c:
```diff
@@ -0,0 +1,102 @@
+#include "wasi.h"
+#include <pthread.h>
+
+// copied from internal/pthread_impl.h
```
Where was this copied from exactly - what is the license on the file that this copied from?

yowl replied:
https://github.com/WebAssembly/wasi-libc/blob/main/libc-top-half/musl/src/internal/pthread_impl.h

There's no header on that file, but https://github.com/WebAssembly/wasi-libc/tree/1dfe5c302d1c5ab621f7abf04620fae92700fd22 has various license files.  As this is `libc-top-half/musl` it looks like it is MIT

https://github.com/WebAssembly/wasi-libc/blob/1dfe5c302d1c5ab621f7abf04620fae92700fd22/LICENSE#LL12C7-L12C72

and

https://github.com/WebAssembly/wasi-libc/blob/1dfe5c302d1c5ab621f7abf04620fae92700fd22/libc-top-half/musl/COPYRIGHT#LL1C1-L1C1

jkotas commented on src/coreclr/nativeaot/Runtime/wasm/wasi_pthread.c:
```diff
@@ -0,0 +1,102 @@
+#include "wasi.h"
+#include <pthread.h>
+
+// copied from internal/pthread_impl.h
```
```suggestion
// copied from https://github.com/WebAssembly/wasi-libc/blob/main/libc-top-half/musl/src/internal/pthread_impl.h
```

yowl replied:
Thanks, I will do that in future.

jkotas committed:
Update src/coreclr/nativeaot/Runtime/wasm/wasi_pthread.c

jkotas commented:
Thank you, @yowl and @SingleAccretion !
